src/uct/tcp/tcp_ep.c

```diff
@@ -537,20 +530,67 @@ void uct_tcp_ep_pending_queue_dispatch(uct_tcp_ep_t *ep)
     }
 }
 
-static void uct_tcp_ep_handle_disconnected(uct_tcp_ep_t *ep,
-                                           uct_tcp_ep_ctx_t *ctx)
+static UCS_F_ALWAYS_INLINE void uct_tcp_ep_advance_tx(uct_tcp_ep_t *ep,
+                                                      const uct_tcp_am_hdr_t *hdr)
 {
-    ucs_debug("tcp_ep %p: remote disconnected", ep);
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+
+    ep->tx.length      += sizeof(*hdr) + hdr->length;
+    iface->outstanding += ep->tx.length;
+}
+
+static UCS_F_ALWAYS_INLINE void uct_tcp_ep_comp_tx(uct_tcp_ep_t *ep,
+                                                   size_t sent_length)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+
+    iface->outstanding -= sent_length;
+    ep->tx.offset      += sent_length;
+}
 
-    uct_tcp_ep_ctx_reset(ctx);
+static UCS_F_ALWAYS_INLINE void uct_tcp_ep_comp_zcopy(uct_tcp_ep_t *ep,
+                                                      uct_completion_t *comp,
+                                                      ucs_status_t status)
+{
+    ep->ctx_caps &= ~UCS_BIT(UCT_TCP_EP_CTX_TYPE_ZCOPY_TX);
```

Comment:
does it mean that ep can have only 1 outstanding zcopy op?

Reply:
yes, TCP can have on;y 1 outstanding Zcopy, since it couldn't send many buffers thru a single socket